rpc : reuse compute graphs

rgerganov opened the pull request:
Store compute graphs on the server side and reuse them when possible. ~~Compute graphs are kept in a ring buffer with fixed size, so we can avoid serializing and deserializing the same graph every time. Add two new commands:~~
 ~~* RPC_CMD_GRAPH_COMPUTE_AND_STORE -- compute and store the graph~~
 ~~* RPC_CMD_GRAPH_RECOMPUTE -- recompute the graph with the given ID~~

~~Currently there is no good way to associate an ID with `ggml_cgraph`, so we abuse `tensor->extra` of the first node for this purpose.~~

jukofyork commented:
When I'm running over RPC, I'm noticing the amount of data sent from the main host to the RPC servers is very asymmetric during token generation, eg:

- Something like 200MB/s getting sent to each RPC server.
- Something like 5MB/s getting sent from each RPC server.

Is this the compute graph getting sent (which I can only think must have all the sin/cos values for the positional encodings or something to have this large disparity?) and the hidden state(s) getting sent back?

rgerganov committed:
rpc : cache and reuse compute graphs

Store the last computed graph and reuse it when possible.
Also do not return response from GRAPH_COMPUTE and assume it always
completes successfully. If this this is not the case, the server closes
the connection. This saves us a network round trip to the server.

jukofyork commented:
I think I've deciphered this now (it was confusing because I was also running speculative decoding!).

> For graph_compute, the received data in bytes is n_nodes*8 + n_tensors*296

```
[graph_compute] device: 0, n_nodes: 1080, n_tensors: 1400
```

So this is `1080*8 + 1400*296 = 423040`.

---

```
[set_tensor] buffer: 0x561d64889540, data: 0x7f7c8a000000, offset: 0, size: 20480
```

this is the hidden state stored as `float32` (ie: `20480/4 = 5120`).

---

```
[set_tensor] buffer: 0x561d64889540, data: 0x7f7c8a05f280, offset: 0, size: 131072
```

This appears to be 2x the context length, so I assume it is something holding position encodings.

---

So overall we are sending `423040+131072+20480` and receiving `20480` back here:

```
[get_tensor] buffer: 0x561d642c72c0, data: 0x7f7c4e0be000, offset: 0, size: 20480
```

which is a `28:1` ratio and in line with the measurements.

rgerganov commented:
The size of the compute graph is substantial in your case (423040 bytes), so reusing the graph (what this PR is about) could bring noticeable improvement for you. I will rebase this PR, so you can test in your environment.

jukofyork commented:
> The size of the compute graph is substantial in your case (423040 bytes), so reusing the graph (what this PR is about) could bring noticeable improvement for you. I will rebase this PR, so you can test in your environment.

Thanks!

jukofyork commented:
> > > ```c++
> > >     if (!send_data(sock->fd, &cmd_byte, sizeof(cmd_byte))) {
> > >         return false;
> > >     }
> > >     if (!send_data(sock->fd, &input_size, sizeof(input_size))) {
> > >         return false;
> > >     }
> > >     if (!send_data(sock->fd, input, input_size)) {
> > >         return false;
> > >     }
> > > ```
> > 
> > 
> > With `TCP_NODELAY`, each of these calls may be generating a new packet, which may add to the network latency. It may be worth building the entire packet first in a buffer and sending it with a single call to `send`.
> 
> I tried this a few days ago but it didn't make any difference for me. Not sure if it's the extra `memcpy` required to place the 9 bytes before the data or if it's just my network stack is still buffering small packets behind the scenes with `TCP_NODELAY` turned off.

Here's what I tried for both `send_msg` and `send_rpc_cmd`, but it didn't seem to help me at all:

https://github.com/ggml-org/llama.cpp/commit/5ce87d1316bb744ca320fddb5520a7ca821e09ed

```cpp
static bool send_msg(sockfd_t sockfd, const void * msg, size_t msg_size) {
    const size_t header_size = sizeof(msg_size);
    std::vector<uint8_t> buf;
    buf.resize(header_size + msg_size);

    // header
    memcpy(buf.data(), &msg_size, sizeof(msg_size));

    // payload
    if (msg_size > 0) {
        memcpy(buf.data() + header_size, msg, msg_size);
    }

    // single send
    return send_data(sockfd, buf.data(), buf.size());
}
```

```cpp
static bool send_rpc_cmd(const std::shared_ptr<socket_t> & sock, enum rpc_cmd cmd, const void * input, size_t input_size) {
    const size_t header_size = 1 + sizeof(input_size);
    std::vector<uint8_t> buf;
    buf.resize(header_size + input_size);

    // header
    buf[0] = static_cast<uint8_t>(cmd);
    memcpy(buf.data() + 1, &input_size, sizeof(input_size));

    // payload
    if (input_size > 0) {
        memcpy(buf.data() + header_size, input, input_size);
    }

    // single send (send_data may still chunk very large buffers, which is fine)
    return send_data(sock->fd, buf.data(), buf.size());
}
```

jukofyork commented:
> > > > ```c++
> > > >     if (!send_data(sock->fd, &cmd_byte, sizeof(cmd_byte))) {
> > > >         return false;
> > > >     }
> > > >     if (!send_data(sock->fd, &input_size, sizeof(input_size))) {
> > > >         return false;
> > > >     }
> > > >     if (!send_data(sock->fd, input, input_size)) {
> > > >         return false;
> > > >     }
> > > > ```
> > > 
> > > 
> > > With `TCP_NODELAY`, each of these calls may be generating a new packet, which may add to the network latency. It may be worth building the entire packet first in a buffer and sending it with a single call to `send`.
> > 
> > 
> > I tried this a few days ago but it didn't make any difference for me. Not sure if it's the extra `memcpy` required to place the 9 bytes before the data or if it's just my network stack is still buffering small packets behind the scenes with `TCP_NODELAY` turned off.
> 
> Here's what I tried for both `send_msg` and `send_rpc_cmd`, but it didn't seem to help me at all:
> 
> [5ce87d1](https://github.com/ggml-org/llama.cpp/commit/5ce87d1316bb744ca320fddb5520a7ca821e09ed)
> 
> ```c++
> static bool send_msg(sockfd_t sockfd, const void * msg, size_t msg_size) {
>     const size_t header_size = sizeof(msg_size);
>     std::vector<uint8_t> buf;
>     buf.resize(header_size + msg_size);
> 
>     // header
>     memcpy(buf.data(), &msg_size, sizeof(msg_size));
> 
>     // payload
>     if (msg_size > 0) {
>         memcpy(buf.data() + header_size, msg, msg_size);
>     }
> 
>     // single send
>     return send_data(sockfd, buf.data(), buf.size());
> }
> ```
> 
> ```c++
> static bool send_rpc_cmd(const std::shared_ptr<socket_t> & sock, enum rpc_cmd cmd, const void * input, size_t input_size) {
>     const size_t header_size = 1 + sizeof(input_size);
>     std::vector<uint8_t> buf;
>     buf.resize(header_size + input_size);
> 
>     // header
>     buf[0] = static_cast<uint8_t>(cmd);
>     memcpy(buf.data() + 1, &input_size, sizeof(input_size));
> 
>     // payload
>     if (input_size > 0) {
>         memcpy(buf.data() + header_size, input, input_size);
>     }
> 
>     // single send (send_data may still chunk very large buffers, which is fine)
>     return send_data(sock->fd, buf.data(), buf.size());
> }
> ```

Actually this now seems to be giving me quite a boost! With this PR and that branch merged, I'm now getting this:

| model                          |       size |     params | backend    | ngl | fa | dev          | ts           |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | -: | ------------ | ------------ | --------------: | -------------------: |
| glm4moe 355B.A32B Q6_K         | 230.78 GiB |   356.79 B | CUDA,RPC   |  99 |  1 | CUDA0/RPC0/RPC1/RPC2/RPC3/CUDA1 | 16.00/15.00/15.00/15.00/15.00/17.00 |           pp512 |        354.63 ± 1.07 |
| glm4moe 355B.A32B Q6_K         | 230.78 GiB |   356.79 B | CUDA,RPC   |  99 |  1 | CUDA0/RPC0/RPC1/RPC2/RPC3/CUDA1 | 16.00/15.00/15.00/15.00/15.00/17.00 |           tg128 |         22.70 ± 0.01 |
| glm4moe 355B.A32B Q6_K         | 230.78 GiB |   356.79 B | CUDA,RPC   |  99 |  1 | CUDA0/RPC0/RPC1/RPC2/RPC3/CUDA1 | 16.00/15.00/15.00/15.00/15.00/17.00 |           tg256 |         22.41 ± 0.01 |
| glm4moe 355B.A32B Q6_K         | 230.78 GiB |   356.79 B | CUDA,RPC   |  99 |  1 | CUDA0/RPC0/RPC1/RPC2/RPC3/CUDA1 | 16.00/15.00/15.00/15.00/15.00/17.00 |           tg512 |         22.07 ± 0.05 |

compared to this the other day:

> This makes quite a big difference for me:
> ## master
> model 	size 	params 	backend 	ngl 	fa 	dev 	ts 	test 	t/s
> glm4moe 355B.A32B Q6_K 	231.00 GiB 	356.79 B 	CUDA,RPC 	99 	1 	CUDA0/RPC0/RPC1/RPC2/RPC3/CUDA1 	16.00/15.00/15.00/15.00/15.00/17.00 	pp512 	318.55 ± 0.86
> glm4moe 355B.A32B Q6_K 	231.00 GiB 	356.79 B 	CUDA,RPC 	99 	1 	CUDA0/RPC0/RPC1/RPC2/RPC3/CUDA1 	16.00/15.00/15.00/15.00/15.00/17.00 	tg128 	15.64 ± 0.00
> glm4moe 355B.A32B Q6_K 	231.00 GiB 	356.79 B 	CUDA,RPC 	99 	1 	CUDA0/RPC0/RPC1/RPC2/RPC3/CUDA1 	16.00/15.00/15.00/15.00/15.00/17.00 	tg256 	15.64 ± 0.01
> glm4moe 355B.A32B Q6_K 	231.00 GiB 	356.79 B 	CUDA,RPC 	99 	1 	CUDA0/RPC0/RPC1/RPC2/RPC3/CUDA1 	16.00/15.00/15.00/15.00/15.00/17.00 	tg512 	15.54 ± 0.01
> ## PR
> model 	size 	params 	backend 	ngl 	fa 	dev 	ts 	test 	t/s
> glm4moe 355B.A32B Q6_K 	231.00 GiB 	356.79 B 	CUDA,RPC 	99 	1 	CUDA0/RPC0/RPC1/RPC2/RPC3/CUDA1 	16.00/15.00/15.00/15.00/15.00/17.00 	pp512 	320.18 ± 0.99
> glm4moe 355B.A32B Q6_K 	231.00 GiB 	356.79 B 	CUDA,RPC 	99 	1 	CUDA0/RPC0/RPC1/RPC2/RPC3/CUDA1 	16.00/15.00/15.00/15.00/15.00/17.00 	tg128 	19.47 ± 0.02
> glm4moe 355B.A32B Q6_K 	231.00 GiB 	356.79 B 	CUDA,RPC 	99 	1 	CUDA0/RPC0/RPC1/RPC2/RPC3/CUDA1 	16.00/15.00/15.00/15.00/15.00/17.00 	tg256 	19.48 ± 0.01
> glm4moe 355B.A32B Q6_K 	231.00 GiB 	356.79 B 	CUDA,RPC 	99 	1 	CUDA0/RPC0/RPC1/RPC2/RPC3/CUDA1 	16.00/15.00/15.00/15.00/15.00/17.00 	tg512 	19.30 ± 0.01
> 
> `(19.47 + 19.48 + 19.30) / (15.64 + 15.64 + 15.54)` = `+24.5%` tg improvement.

It may be something else that has changed, so will check next week to be sure, but it seems odd I got around 15 tokens/s when I tested this branch a couple of days ago.

jukofyork commented:
I made a draft PR https://github.com/ggml-org/llama.cpp/pull/16892 to see if this makes any difference to others, or if it only helps with this specific PR, etc.

kyuz0 commented:
On my setup (two AMD Strix Halo hosts), this PR has improved (~5.5%) prompt processing:

```
  Device 0: AMD Radeon Graphics, gfx1151 (0x1151), VMM: no, Wave Size: 32
| model                          |       size |     params | backend    | ngl | fa | mmap |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | -: | ---: | --------------: | -------------------: |
| minimax-m2 230B.A10B Q6_K      | 180.94 GiB |   228.69 B | ROCm,RPC   |  99 |  1 |    0 |           pp512 |        175.56 ± 1.40 |
| minimax-m2 230B.A10B Q6_K      | 180.94 GiB |   228.69 B | ROCm,RPC   |  99 |  1 |    0 |           tg128 |         17.50 ± 0.08 |
```

Right now I only have a screenshot of the performance without this PR:

<img width="2880" height="430" alt="image" src="https://github.com/user-attachments/assets/c61b4211-8a8a-4352-9e2d-0cca824e74e2" />

Happy to do more testing. My setup currenlt uses USB-to-USB, 10Gbps and I'm trying to see if I can get to 20-40Gbps.

rgerganov commented:
I have reimplemented the graph caching using the approach suggested by @ggerganov. Both client and server maintain a ring buffer of cached compute graphs and reuse them when possible. The newly added RPC commands for computing graphs do not return result -- if `graph_compute()` returns something different from `GGML_STATUS_SUCCESS`, the server simply closes the connection. This will save a network round-trip and should bring noticeable TG speed improvement.

@kyuz0 @jukofyork Could you please test this PR on your testbeds and report the results here?

rgerganov commented:
I am seeing 13% TG improvement with gpt-oss-20b when running on localhost:

## master

| model                          |       size |     params | backend    | ngl |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | --------------: | -------------------: |
| gpt-oss 20B MXFP4 MoE          |  11.27 GiB |    20.91 B | RPC        |  99 |           pp512 |     9332.25 ± 120.10 |
| gpt-oss 20B MXFP4 MoE          |  11.27 GiB |    20.91 B | RPC        |  99 |           tg128 |        251.03 ± 1.38 |


## PR

### 

| model                          |       size |     params | backend    | ngl |            test |                  t/s |
| ------------------------------ | ---------: | ---------: | ---------- | --: | --------------: | -------------------: |
| gpt-oss 20B MXFP4 MoE          |  11.27 GiB |    20.91 B | RPC        |  99 |           pp512 |     9473.22 ± 108.71 |
| gpt-oss 20B MXFP4 MoE          |  11.27 GiB |    20.91 B | RPC        |  99 |           tg128 |        283.68 ± 0.83 |

eugr commented:
I'm seeing a bit of improvement with dual DGX Sparks connected via their 200G port on Qwen3-VL-235B (and even GPT-OSS-120B), but for some reason not on MiniMax M2.

# Qwen3-235B

## Before

| model                              |       size |   params | backend  |            test |            t/s |
| ---------------------------------- | ---------: | -------: | -------- | --------------: | -------------: |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC |          pp2048 |  528.52 ± 2.88 |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC |            tg32 |   12.98 ± 0.05 |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC |  pp2048 @ d4096 |  469.70 ± 5.47 |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC |    tg32 @ d4096 |   11.62 ± 0.08 |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC |  pp2048 @ d8192 |  420.87 ± 8.01 |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC |    tg32 @ d8192 |   11.15 ± 0.08 |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC | pp2048 @ d16384 |  340.40 ± 8.40 |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC |   tg32 @ d16384 |    9.90 ± 0.02 |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC | pp2048 @ d32768 | 226.70 ± 35.69 |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC |   tg32 @ d32768 |    8.03 ± 0.04 |

## After

| model                              |       size |   params | backend  |            test |            t/s |
| ---------------------------------- | ---------: | -------: | -------- | --------------: | -------------: |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC |          pp2048 |  551.75 ± 1.11 |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC |            tg32 |   13.78 ± 0.02 |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC |  pp2048 @ d4096 | 491.34 ± 22.40 |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC |    tg32 @ d4096 |   12.47 ± 0.09 |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC |  pp2048 @ d8192 |  454.67 ± 1.51 |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC |    tg32 @ d8192 |   12.15 ± 0.02 |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC | pp2048 @ d16384 |  377.73 ± 0.39 |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC |   tg32 @ d16384 |   10.91 ± 0.02 |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC | pp2048 @ d32768 |  279.02 ± 0.82 |
| qwen3vlmoe 235B.A22B Q4_K - Medium | 124.91 GiB | 235.09 B | CUDA,RPC |   tg32 @ d32768 |    8.99 ± 0.07 |

# MiniMax M2

## Before

| model                              |       size |   params | backend  |            test |           t/s |
| ---------------------------------- | ---------: | -------: | -------- | --------------: | ------------: |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC |          pp2048 | 906.42 ± 1.27 |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC |            tg32 |  25.32 ± 0.28 |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC |  pp2048 @ d4096 | 822.09 ± 4.14 |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC |    tg32 @ d4096 |  21.47 ± 0.16 |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC |  pp2048 @ d8192 | 736.49 ± 6.00 |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC |    tg32 @ d8192 |  19.03 ± 0.12 |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC | pp2048 @ d16384 | 615.61 ± 5.00 |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC |   tg32 @ d16384 |  15.49 ± 0.22 |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC | pp2048 @ d32768 | 460.02 ± 5.47 |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC |   tg32 @ d32768 |  11.14 ± 0.07 |

## After

| model                              |       size |   params | backend  |            test |           t/s |
| ---------------------------------- | ---------: | -------: | -------- | --------------: | ------------: |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC |          pp2048 | 901.43 ± 2.48 |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC |            tg32 |  25.90 ± 0.08 |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC |  pp2048 @ d4096 | 810.92 ± 1.31 |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC |    tg32 @ d4096 |  23.10 ± 0.27 |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC |  pp2048 @ d8192 | 731.97 ± 3.02 |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC |    tg32 @ d8192 |  19.81 ± 0.14 |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC | pp2048 @ d16384 | 606.83 ± 1.97 |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC |   tg32 @ d16384 |  16.08 ± 0.08 |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC | pp2048 @ d32768 | 456.50 ± 1.00 |
| minimax-m2 230B.A10B Q4_K - Medium | 122.58 GiB | 228.69 B | CUDA,RPC |   tg32 @ d32768 |  11.71 ± 0.07 |

ggerganov commented on ggml/src/ggml-rpc/ggml-rpc.cpp:
```diff
@@ -106,13 +106,16 @@ enum rpc_cmd {
 // Try RPC_CMD_SET_TENSOR_HASH first when data size is larger than this threshold
-const size_t HASH_THRESHOLD = 10 * 1024 * 1024;
+const size_t   HASH_THRESHOLD    = 10 * 1024 * 1024;
+const int      MAX_STORED_GRAPHS = 64;
```
Not sure if storing so many graphs is not an overkill. I think just keeping the last graph should be enough.

rgerganov commented on ggml/src/ggml-rpc/ggml-rpc.cpp:
```diff
@@ -876,11 +943,23 @@ ggml_backend_buffer_type_t ggml_backend_rpc_buffer_type(const char * endpoint, u
+    } else {
+        gc = new graph_cache;
+        cache_map[endpoint] = gc;
+    }
```
When the `graph_cache` is shared for the entire endpoint, we can easily keep `curr_graph_id` in sync between client and server. I thought this is simpler compared to maintaining a separate `graph_cache` and `curr_graph_id` for each device on the endpoint. On second thought, having one `graph_cache` per context seems better, so I can revisit this.

kyuz0 commented:
> > https://github.com/kyuz0/amd-strix-halo-toolboxes/blob/main/toolboxes/Dockerfile.rocm-6.4.4-rocwmma
> 
> on fedora:43 rocm 6.4.4 is now upstream, no need to use the "el9" .
> 
> in my case I create a toolbox (I have silverblue ;)):
> 
> ```
> toolbox create -d fedora -r 43 rocm-devel-fc43
> toolbox enter rocm-devel-fc43
> 
> # add rocm+dep:
>  sudo dnf upgrade --refresh
>  sudo dnf install hipblas-devel rocm-hip-devel rocblas-devel
>  sudo dnf install cmake gcc-c++
>  sudo dnf install rocm-smi rocminfo amd-smi
>  sudo dnf install libcurl-devel
> 
> # build with
>  cmake -S . -B build/rocm6.4.4 -DGGML_HIP=ON -DGPU_TARGETS=gfx1151 -DCMAKE_BUILD_TYPE=Release
>  cmake --build build/rocm6.4.4 --config Release -- -j 16
> 
> # ...
> ```

Thank you, I still think regardless of where ROCm comes from, I am not sure why I am getting the same performance with and without the PR, and again, it seems even without the PR now performance is better. 

I suspect user error, so I might just create a toolbox manually instead of my pipelines.
